src/main/java/bio/terra/service/resourcemanagement/google/GoogleResourceService.java

```diff
@@ -115,38 +116,131 @@ public GoogleBucketResource getBucketResourceById(UUID bucketResourceId, boolean
 
     /**
      * Fetch/create a bucket cloud resource and the associated metadata in the bucket_resource table.
-     * The flightId is used to lock the bucket metadata during possible creation.
-     * Below is an outline of the logic flow.
-     *   1. Try to insert a new bucket_resource row and lock it
-     *      If the insert failed, lookup the existing row
-     *   2. Check who holds the lock on the bucket_resource row
-     *      a. GET case = row is unlocked. bucket has already been created
-     *         Check that the bucket cloud resource exists, throw an exception if not.
-     *      b. CREATE case = row is locked by this flight. need to create the bucket here
-     *         Check if the bucket cloud resource exists.
-     *           i. If not, create it
-     *           ii. If it does exist and this is a testing environment, just use it
-     *           iii. If it does exist and this is a recovery flight, just use it
-     *           iv. Otherwise, the state of the cloud is out of sync with the metadat
-     *               Delete anything we've created and throw an exception
-     *      c. LOCKED case = the row is locked by another flight. throw a lock exception so the flight knows to retry
-     * @param bucketRequest
-     * @param flightId
+     *
+     * On entry to this method, there are 9 states along 3 main dimensions:
+     * Google Bucket - exists or not
+     * DR Metadata record - exists or not
+     * DR Metadata lock state (only if record exists):
+     *  - not locked
+     *  - locked by this flight
+     *  - locked by another flight
+     * In addition, there is one case where it matters if we are reusing buckets or not.
+     *
+     * Itemizing the 9 cases:
+     * CASE 1: bucket exists, record exists, record is unlocked
+     *   The predominant case. We return the bucket resource
+     *
+     * CASE 2: bucket exists, record exists, locked by another flight
+     *   We have to wait until the other flight finishes creating the bucket. Throw BucketLockFailureException.
+     *   We expect the calling Step to retry on that exception.
+     *
+     * CASE 3: bucket exists, record exists, locked by us
+     *   This flight created the bucket, but failed before we could unlock it. So, we unlock and
+     *   return the bucket resource.
+     *
+     * CASE 4: bucket exists, no record exists, we are allowed to reuse buckets
+     *   This is a common case in development where we re-use the same cloud resources over and over during
+     *   testing rather than continually create and destroy them. In this case, we proceed with the
+     *   try-to-create-bucket algorithm.
```

Comment:
Perhaps change to "try-to-create-bucket-metadata" algorithm.

Reply:
Sounds good. These changes will show up in PR #458